torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -0,0 +1,192 @@
+# and return nothing (this is a common class of operators).
+# torch.compile will automatically support this op without anyone needing
+# to provide a functionalization kernel for it. Here's how.
+#
+# Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> None
+# op. First, when FakeTensor sees this op, it can just return None.
+# This is the case because custom ops are not allowed to mutate input
+# metadata, so the FakeTensor rule for one that returns None is trivial.
+#
+# Next, when Python FunctionalTensor sees the op, it will functionalize
+# it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
+# HOP and replacing the mutated inputs with the outputs of this HOP.
+# This HOP effectively runs the functional version of the op when
+# called: it clones inputs that will be mutated, runs the op, and
+# then returns Tensors with the new values.
+
+
+class AutoFunctionalized(HigherOrderOperator):
+    """auto_functionalized(op, mutated_args_names, kwargs)
+
+    This HOP runs a "functional" version of op.
+
+    Concretely, it clones kwargs that `op` mutates (specified by
+    mutated_args_names), runs `op(**kwargs)` with the cloned values,
+    and then returns a flat Tuple of the cloned values that were mutated.
+
+    We have some restrictions on `op`, most notably that it returns None.
+    See `can_auto_functionalize` for the restrictions. We can likely lift
+    many of these if users request it.
+    """
+
+    def __init__(self):
+        super().__init__("auto_functionalized")
+
+    def __call__(
+        self,
+        op: torch._ops.OpOverload,
+        mutated_args_names: List[str],
+        kwargs: Dict[str, Any],
+    ) -> Tuple[Tensor, ...]:
+        assert can_auto_functionalize(op)
+        assert isinstance(mutated_args_names, list)
+        assert isinstance(kwargs, dict)
+        return super().__call__(op, mutated_args_names, kwargs)
+
+
+auto_functionalized = AutoFunctionalized()
+
+
+def can_auto_functionalize(op: torch._ops.OperatorBase) -> bool:
+    if not isinstance(op, torch._ops.OpOverload):
+        return False
+    schema = op._schema
+    if not schema.is_mutable:
+        return False
+    if len(schema.returns) > 0:
+        return False
+    for arg in schema.arguments:
+        if arg.alias_info is None:
+            continue
+        if not arg.alias_info.is_write:
+            continue
+        if type(arg.type) is torch.TensorType:
+            continue
+        if str(arg.type) == "Optional[Tensor]":
+            continue
+        # Not yet supported: other Tensor types. This includes things like
+        # Tensor[], Tensor?[], Tensor[]?.
+        return False
+    return True
+
+
+@auto_functionalized.py_impl(DispatchKey.CompositeExplicitAutograd)
+def auto_functionalized_dense(
+    op: torch._ops.OpOverload, mutated_args_names: List[str], kwargs: Dict[str, Any]
+) -> Tuple[Tensor, ...]:
+    new_kwargs = dict(**kwargs)
+    result = []
+    for name in mutated_args_names:
```

Comment:
Do you think it's worth a `warn_once()` here as a perf warning? (counterargument == too noisy, and it's not very actionable)

Reply:
Probably not for now, as you mentioned that it is not very actionable. Maybe we can do this in the future when the re-inplacing pass exists (and we can log instead of warn so someone staring at the log might be able to surface this to us)